python/packages/core/agent_framework/_mcp.py

```diff
@@ -883,6 +887,10 @@ def get_mcp_client(self) -> _AsyncGeneratorContextManager[Any, None]:
             args["terminate_on_close"] = self.terminate_on_close
         if self._client_kwargs:
             args.update(self._client_kwargs)
+        if self._include_traceparent:
```

Comment:
I think it’s better to build this around per-request dynamic headers, not mutate the tool's static dict. Capture the static headers once in `MCPStreamableHTTPTool`, then pass an optional `dynamic_headers` callable down into `streamablehttp_client` / `StreamableHTTPTransport`. The callable can pull the current span, make sure the context is valid, derive the sampled flag, and return `{"traceparent": ...}` only when it should be sent. We can then have `_prepare_request_headers` merge that callable's output on every POST so each RPC carries the active span while the static headers stay untouched. That then handles a possible `KeyError` when no headers were supplied and keeps the trace flag accurate instead of hardcoding `01`.